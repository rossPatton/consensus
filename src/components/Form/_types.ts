export type tState = {
  captcha: string,
  error: string,
  hasMounted: boolean,
};

export type tProps = {
  autoComplete?: string, // auto browser autocomplete or not
  className?: string, // optional classname
  encType?: 'multipart/form-data',
  error?: string | ts.fetchResponse<Error>,
  includeCaptcha?: boolean,
  legend: React.ReactNode, // necessary for a11y
  name: string, // necessary unique name to associate with the form

  // client-side js submit (action handles server only submit)
  onSubmit: (captchaToken: string) => void,
  renderFields: (state: {captcha: string, hasMounted: boolean}) => any,
  renderSubmit: (state: {captcha: string, hasMounted: boolean}) => any,
};
