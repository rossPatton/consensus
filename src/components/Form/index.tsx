import HCaptcha from '@hcaptcha/react-hcaptcha';
import cx from 'classnames';
import loglevel from 'loglevel';
import React, { ErrorInfo, PureComponent } from 'react';

import { tProps, tState } from './_types';

class FormContainer extends PureComponent<tProps, tState> {
  static defaultProps = {
    autoComplete: 'off',
    className: '',
    includeCaptcha: false,
  };

  state = {
    error: '',
    hasMounted: false,
    captcha: '',
  };

  componentDidCatch(error: Error, info: ErrorInfo) {
    this.setState({
      error: error.message,
    }, () => loglevel.error(error, info));
    // TODO log to separate service, but only show user 500 page
  }

  componentDidMount() {
    // we do this so we only disable default form submit when js is available
    this.setState({
      hasMounted: true,
    });
  }

  handleVerificationSuccess = (captcha: string) => {
    this.setState({
      captcha,
    });
  }

  onSubmit = (ev: React.FormEvent<HTMLFormElement>) => {
    ev.preventDefault();
    this.props.onSubmit(this.state.captcha);
  }

  render() {
    const {error: errorProp, includeCaptcha, legend} = this.props;
    const {captcha, error: errorState, hasMounted} = this.state;
    let error = errorProp || errorState;

    // strip out status code if included
    if (typeof error === 'string' && error.includes(':')) {
      error = error.split(':')[1];
    }

    const validError = typeof error === 'string' && !!error;

    return (
      <form
        method="post"
        encType={this.props.encType}
        className={this.props.className}
        name={this.props.name}
        autoComplete={this.props.autoComplete}
        onSubmit={this.onSubmit}
      >
        <fieldset>
          <legend>
            {typeof legend === 'string'
              ? (
                <h1 className="mb-2">
                  {legend}
                </h1>
              )
              : legend}
          </legend>
          {this.props.renderFields({captcha, hasMounted})}
          <div
            className={cx({
              'mb-2': includeCaptcha || validError,
            })}>
            {includeCaptcha && (
              <div id="hcaptcha-wrapper" className="mb-2">
                <HCaptcha
                  sitekey={__HCAPTCHA_KEY__}
                  onVerify={this.handleVerificationSuccess}
                />
              </div>
            )}
            {this.props.renderSubmit({captcha, hasMounted})}
          </div>
          {validError && (
            <div className="rounded text-3 bg-red-2 p-2 text-white font-semibold w-full">
              {error}
            </div>
          )}
        </fieldset>
      </form>
    );
  }
}

export default FormContainer;
