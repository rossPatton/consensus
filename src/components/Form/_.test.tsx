import React from 'react';
import render from 'react-test-renderer';

import Form from '.';

// @TODO add checks for all the variants
describe('components/Form', () => {
  it('renders', () => {
    const component = render.create((
      <Form
        encType="multipart/form-data"
        error="Something is wrong"
        name="planEventForm"
        onSubmit={jest.fn()}
        legend={(<h2 className="text-3 mb-1">Legend here</h2>)}
        renderFields={() => (<></>)}
        renderSubmit={() => (<></>)}
      />
    ));

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with server error', () => {
    const component = render.create((
      <Form
        encType="multipart/form-data"
        error="400:Bad Request"
        name="planEventForm"
        onSubmit={jest.fn()}
        legend={(<h2 className="text-3 mb-1">Legend here</h2>)}
        renderFields={() => (<></>)}
        renderSubmit={() => (<></>)}
      />
    ));

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
