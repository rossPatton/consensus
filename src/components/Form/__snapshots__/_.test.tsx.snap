// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Form renders 1`] = `
<form
  autoComplete="off"
  className=""
  encType="multipart/form-data"
  method="post"
  name="planEventForm"
  onSubmit={[Function]}
>
  <fieldset>
    <legend>
      <h2
        className="text-3 mb-1"
      >
        Legend here
      </h2>
    </legend>
    <div
      className="mb-2"
    />
    <div
      className="rounded text-3 bg-red-2 p-2 text-white font-semibold w-full"
    >
      Something is wrong
    </div>
  </fieldset>
</form>
`;

exports[`components/Form renders with server error 1`] = `
<form
  autoComplete="off"
  className=""
  encType="multipart/form-data"
  method="post"
  name="planEventForm"
  onSubmit={[Function]}
>
  <fieldset>
    <legend>
      <h2
        className="text-3 mb-1"
      >
        Legend here
      </h2>
    </legend>
    <div
      className="mb-2"
    />
    <div
      className="rounded text-3 bg-red-2 p-2 text-white font-semibold w-full"
    >
      Bad Request
    </div>
  </fieldset>
</form>
`;
